04-onion-routing.md

```diff
@@ -635,6 +637,35 @@ Alternatively, implementations may choose to apply non-strict forwarding only to
 like-policy channels to ensure their expected fee revenue does not deviate by
 using an alternate channel.
 
+## Resource Bucketing
+
+When making the decision to forward a payment on its outgoing channel, a node 
+MAY choose to limit its exposure to HTLCs that put it at risk of a denial of 
+service attack.
+* `unknown_allocation_slots`: defines the number of HTLC slots allocated to 
+   unknown traffic (default: 50% of remote peer's `max_accepted_htlcs`).
+* `unknown_allocation_liquidity`: defines the amount of the channel balance 
+   that is allocated to unknown traffic (default: 50% of remote peer's 
+   `max_htlc_value_in_flight_msat`.
+
+A node implementing resource bucketing limits exposure on its outgoing channel:
+- MUST choose `unknown_allocation_slots` <= the remote channel peer's 
+  `max_accepted_htlcs`.
+- MUST choose `unknown_allocation_liquidity` <= the remote channel peer's 
+  `max_htlc_value_in_flight_msat`.
+- If `endorsed` is set to 1 in the incoming `update_add_htlc` AND the HTLC 
+  is from a node that the forwarding node considers to have good local 
+  reputation (see [Recommendations for Reputation](#recommendations-for-reputation)):
+  - SHOULD proceed with forwarding the HTLC.
+  - SHOULD set `endorsed` to 1 in the outgoing `update_add_htlc`.
```

Comment:
Not sure if this is the right place to comment, but we could consider temporarily subtracting the `fees` (paid to us) of the HTLC we just endorsed from the `reputation_revenue` of that peer.

When the outcome of the HTLC is decided we grant back that temporary damage, then we apply `effective_fees` to `reputation_revenue` as defined in the formula.

This doesn't affect the long-term behavior of the `reputation_revenue` value, but acts as a safety barrier. A peer can now only have as many HTLCs in the `endorsed` slots as their `reputation_revenue` can reflect on.

e.g I made some payments and juuust started satisfying `reputation_revenue >= routing_revenue`. Previously I would be able to create roughly (with the 50% default ratio for unknown allocations) ~240 HTLCs that would get `endorsed` by the receiving peer. These 240 HTLCs can have a gigantic impact not only on my local reputation with my peer, but on my peer and their peer, and so on along the route.

With the temporary subtraction, as soon as 1 of those 240 HTLCs is accepted my local rep will suffer a `fees` temporary loss, perhaps dropping me temporarily back to the `unkown` slots. This way I can only open my "reputation attack vector" as much as my own local reputation allows.

Reply:
A different (perhaps better) approach to this would be temporarily subtracting the **worst-case** `effective_fee` that this HTLC can induce. If paired with an `effective_fee` upper limit (as discussed [here](https://github.com/lightning/bolts/pull/1071#discussion_r1197626027)) then this would create a worst-case-`effective_fee` threshold for outgoing HTLCs to get endorsed.

This makes the `endorsed` playground a bit more restrained and bottlenecks the scale of possible `endorsed` attacks.

If the above sounds like something we could merge to this proposal, then we could also make access to `endorsed` slots a bit easier (either by increasing the reward or by tweaking the windows)

Reply:
Also could make [this](https://github.com/lightning/bolts/pull/1071#discussion_r1197577251) safer for big LN nodes and the way their reputation is affected.

Reply:
> A different (perhaps better) approach to this would be temporarily subtracting the worst-case effective_fee that this HTLC can induce. 

I think this is a really interesting idea! Took it a step forward and also included in-flight HTLCs, so that we can account for a sudden change in behaviour. 

If we define `worst_case = (routing_window/resolution_period) * htlc_fee`, then we can account for the _potential_ damage a peer can do. 

When we get a `offered_htlc` we subtract the worse case damage of all in flight htlcs *and* this new HTLC to assess reputation:
`reputation_revenue = Σ(effective_fees(settled_htlcs)) - Σ(worst_case(in flight htlcs)) - worst_case(offered htlc)`

Reply:
> If we define `worst_case = (routing_window/resolution_period) * htlc_fee`, then we can account for the potential damage a peer can do.

Something that I don't know if I'd identify as an "issue" is that the `effective_fees` in the worst case are by far larger than those in the best case (i.e HTLC succeeded within ~0s).

Pros: this still gives us the gain-slow / lose-fast trait
Cons: you now need to build up even more `reputation_revenue` to make sufficient use of `endorsed` slots

> reputation_revenue = Σ(effective_fees(settled_htlcs)) - Σ(worst_case(in flight htlcs)) - worst_case(offered htlc)

Perhaps replace `worst_case(offered htlc)` with something in the order of magnitude of the best-case reward
e.g consider `10 * resolution_period` or some other multiplier, but keep the actual `worst_case` for **in-flight**.

`worst_case(offered htlc)` can be huge and could put an unrealistically tough barrier for first-time endorsing

Reply:
I think that offering any type of discount probably opens us up for abuse? Pushed [2600b6f](https://github.com/lightning/bolts/pull/1071/commits/2600b6f6822e643703d35c14d83b8cef6a8c36f3) with an example of how we could look at in-flight HTLCs. 

> could put an unrealistically tough barrier for being endorsed

Do you think that'll be the case for in-flight? I was thinking that while you have a few in-flight, your reputation _temporarily_ drops and then rebounds when they resolve - seems like quite a nice property to have since we never know whether a single HTLC will be used to inflict a lot of damage.

Reply:
> I think that offering any type of discount probably opens us up for abuse? Pushed [2600b6f](https://github.com/lightning/bolts/pull/1071/commits/2600b6f6822e643703d35c14d83b8cef6a8c36f3) with an example of how we could look at in-flight HTLCs.

:+1: 

> Do you think that'll be the case for in-flight? I was thinking that while you have a few in-flight, your reputation temporarily drops and then rebounds when they resolve - seems like quite a nice property to have since we never know whether a single HTLC will be used to inflict a lot of damage.

Yeah your rep would temporarily drop (by something that reflects the average "bad case") and then regardless of the outcome it would rebound (and at that point the `effective_fees` would be applied).

> Σ(effective_fees(settled_htlcs)) - Σ(worst_case(in flight htlcs)) - worst_case(offered htlc)

What I'm considering is that `worst_case(offered htlc)` is huge due to the very big `routing_window`, and it would raise the barrier for the switching point between unendorsed/endorsed forwards.

----

Another thing to consider for this in-flight formula is that we actively reduce the potential throughput of the network (while under attack). Previously we would allow usage up to 100% of the endorsed slots without too much caution. Now, by considering the temporary damage, it would require `endorsed_slots * worst_case(fee)` reputation for a single node to utilize 100% of slots simultaneously.

Reply:
> What I'm considering is that worst_case(offered htlc) is huge due to the very big routing_window, and it would raise the barrier for the switching point between unendorsed/endorsed forwards.

Updated to have account for the _actual_ cltv expiry on each HLTC. 

> Previously we would allow usage up to 100% of the endorsed slots without too much caution. Now, by considering the temporary damage, it would require endorsed_slots * worst_case(fee) reputation for a single node to utilize 100% of slots simultaneously.

Interesting question here: how many in-flight (endorsed) HTLCs do we expect to have on a channel at one time? In the honest case, I'd think we should be clearing out pretty quickly so maybe only allowing a few slots at a time is reasonable? Busy looking at some data to figure this out - I think you're right that rn we're _too_ harsh, but I don't know exactly _how_ permissive we actually want to be.

Reply:
Made this [outstanding-risk "calculator"](https://github.com/lightning/bolts/files/11774650/Reputation.-.In.Flight.HTLCs.xlsx) to see what these values look like for various activity levels/timelocks - indeed probably too harsh (very difficult to have more than 1 in flight HTLC).